drivers/adc/adm1177/adm1177.c

```diff
@@ -0,0 +1,316 @@
+
+	switch(device->last_command) {
+	case ADM1177_CURRENT_EN:
+		ret = no_os_i2c_read(device->i2c_desc, data, 2, 1);
+		if(ret)
+			return ret;
+
+		if(conv_current == NULL)
+			return -EINVAL;
+
+		/* First byte received contains the 8 MSBs of the
+		   current and is left shifted by 4, then the second
+		   byte contains the 4 LSBs of the current and is right
+		   shifted by 4, and when the OR operation between the
+		   two bytes happen we obtain the a 12-bit value of the
+		   current. */
+		*conv_current = (data[0] << 4) | (data[1] >> 4);
+
+		return 0;
+	case ADM1177_VOLTAGE_EN:
+		ret = no_os_i2c_read(device->i2c_desc, data, 2, 1);
+		if(ret)
+			return ret;
+
+		if(conv_voltage == NULL)
+			return -EINVAL;
+
+		/* First byte received contains the 8 MSBs of the
+		   voltage and is left shifted by 4, then the second
+		   byte contains the 4 LSBs of the voltage and is right
+		   shifted by 4, and when the OR operation between the
+		   two bytes happen we obtain the a 12-bit value of
+		   the voltage. */
+		*conv_voltage = (data[0] << 4) | (data[1] >> 4);
+
+		return 0;
+	case ADM1177_VOLTAGE_AND_CURRENT_EN:
+		ret = no_os_i2c_read(device->i2c_desc, data, 3, 1);
+		if(ret)
+			return ret;
+
+		/* In the data array we obtain 3 bytes from the read,
+		   first one contains the 8 MSBs of the Voltage value,
+		   second one contains the 8 MSBs of the Current value,
+		   and the third one contains the 4 LSBs of the Voltage
+		   then the Current value, respectively.
+		   For the Voltage - By left shifting the first byte by
+		   4 we are shifting the 8 MSBs of the voltage by 4,
+		   and adding the third byte right shifted by 4 we add
+		   the remaining 4 LSBs of the voltage.
+		   For the Current -  By left shifting the second
+		   byte by 4 we are shifting the 8 MSBs of the current
+		   by 4, and adding the third byte masked by 0xF
+		   we add the remaining 4 LSBs of the current.*/
+		*conv_voltage = (data[0] << 4) | (data[2] >> 4);
+
+		*conv_current = (data[1] << 4) | (data[2] & 0xF);
+
+		return 0;
+	default:
+		return -EINVAL;
+	}
+}
+
+/**************************************************************************//**
+ * @brief - Converts the raw sample from the Readback into uV.
+ *
+ * @param raw_sample - The raw sample of the Voltage Readback from the ADC.
+ *
+ * @param v_fullscale - adm1177_vrange set to ADM1177_VRANGE_LOW
+ * 			(VRANGE bit is 0).
+ *                      adm1177_vrange set to ADM1177_VRANGE_HIGH
+ * 			(VRANGE bit is 1).
+ *
+ * @param voltage_uv - The pointer through which the converted microvolt value
+ * 		       is returned.
+ *
+ * @return - 0 in case of succes.
+ * 	     negative error code if v_fullscale is set wrong.
+******************************************************************************/
+int adm1177_to_microvolts(uint16_t raw_sample,
+			  int v_fullscale,
+			  uint64_t *voltage_uv)
+{
+	switch(v_fullscale) {
+	case ADM1177_VRANGE_LOW:
+		*voltage_uv = (((uint64_t)ADM1177_V_FULLSCALE_0 *
```

Comment:
it would probably make sense that this function has the signature:
```
int adm1177_to_microvolts(dev, raw_in, volts_out)
```

And within the function you use the dev->vrange to perform the computation.

Reply:
Will do.